### PR TITLE
HIVE-26237: Check if replication cause metastore connection leakage

### PR DESCRIPTION
**Pull request by hmangla98**
Problems:
1. Some replication tests don't close the ongoing HMS connections.
2. At one place in ReplLoadTask, we initiate a fresh HMS connection but don't close it.
3. In DDLOperationContext.java, sometimes stored Hive db object is closed and a new object is instantiated. In this case, Actual ddl operations (say CreateDatabaseOperation) will make fresh HMS connection in outdated Hive obj and it will never get closed since that this connection is not accessible. This is causing connection leakage.

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 1]
```diff
@@ -262,17 +263,20 @@ public static void tearDownAfterClass(){
...
-    } catch (IOException e) {
-
-    }
     Path warehousePathReplica = new Path(MetastoreConf.getVar(hconfMirror, MetastoreConf.ConfVars.WAREHOUSE));
     try {
+      warehousePath.getFileSystem(hconf).delete(warehousePath, true);
       warehousePathReplica.getFileSystem(hconfMirror).delete(warehousePathReplica, true);
-    } catch (IOException e) {
+      Hive.getCurrHiveDb(null, null).close(true);
+    } catch (IOException | HiveException e) {
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:272]: should we at least log the exception?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 2]
```diff
@@ -500,6 +500,10 @@ public static Hive getThreadLocal() {
     return hiveDB.get();
   }
 
+  public static Hive getCurrHiveDb(Hive hiveDb, HiveConf c) throws HiveException {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:503]: Do we really want to have a new method for this?

What if we modify the `get`, or the `getThreadLocal` to provide an always open `Hive`?

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/DDLOperationContext.java [hunk 3]
```diff
@@ -51,8 +51,8 @@ public DDLOperationContext(HiveConf conf, Context context, DDLTask task, DDLWork
     this.console = console;
   }
 
-  public Hive getDb() {
-    return db;
+  public Hive getDb() throws HiveException {
+    return Hive.getCurrHiveDb(db, conf);
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/DDLOperationContext.java:55]: Is there a situation when this is `null` or `closed`?

Comment (hmangla98) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/DDLOperationContext.java:55]: Yes, there is a situation where hive connections is closed and it still returns reference of the closed hive object.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java (hunk 1 above)
Comment (hmangla98) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:272]: Noted.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 4]
```diff
@@ -262,17 +263,20 @@ public static void tearDownAfterClass(){
...
-      warehousePath.getFileSystem(hconf).delete(warehousePath, true);
-    } catch (IOException e) {
-
-    }
     Path warehousePathReplica = new Path(MetastoreConf.getVar(hconfMirror, MetastoreConf.ConfVars.WAREHOUSE));
     try {
+      warehousePath.getFileSystem(hconf).delete(warehousePath, true);
       warehousePathReplica.getFileSystem(hconfMirror).delete(warehousePathReplica, true);
-    } catch (IOException e) {
+      Hive.getCurrHiveDb(null, null).close(true);
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:270]: why not just:
````
Hive db = Hive.getThreadLocal().close(true)
````

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java [hunk 5]
```diff
@@ -549,8 +547,7 @@ private void createReplLoadCompleteAckTask() {
         @Override
         public void run() throws SemanticException {
           try {
-            HiveMetaStoreClient metaStoreClient = new HiveMetaStoreClient(conf);
-            long currentNotificationID = metaStoreClient.getCurrentNotificationEventId().getEventId();
+            long currentNotificationID = getHive().getMSC().getCurrentNotificationEventId().getEventId();
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplLoadTask.java:550]: hard to read, split it on few assignments as before

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 6]
```diff
@@ -253,6 +252,7 @@ public class Hive {
   private SynchronizedMetaStoreClient syncMetaStoreClient;
   private UserGroupInformation owner;
   private boolean isAllowClose = true;
+  private boolean isCurrentClosed = false;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:255]: maybe simply `isClosed`?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 7]
```diff
@@ -534,6 +538,7 @@ public static void set(Hive hive) {
   }
 
   public static void closeCurrent() {
+    hiveDB.get().isCurrentClosed = true;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:541]: can we set this flag in:
````
public void close(boolean forceClose) {
    .....
    isClosed = true;
}
````

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java (hunk 4 above)
Comment (hmangla98) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:270]: Done

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java [hunk 8]
```diff
@@ -262,17 +263,20 @@ public static void tearDownAfterClass(){
...
-
-    }
     Path warehousePathReplica = new Path(MetastoreConf.getVar(hconfMirror, MetastoreConf.ConfVars.WAREHOUSE));
     try {
+      warehousePath.getFileSystem(hconf).delete(warehousePath, true);
       warehousePathReplica.getFileSystem(hconfMirror).delete(warehousePathReplica, true);
     } catch (IOException e) {
 
     }
+    Hive.getThreadLocal().close(true);
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java:273]: same as above

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java [hunk 9]
```diff
@@ -121,6 +122,7 @@ static void internalBeforeClassSetupExclusiveReplica(Map<String, String> primary
   public static void classLevelTearDown() throws IOException {
     primary.close();
     replica.close();
+    Hive.getThreadLocal().close(true);
```
[the remark this answers is not in the document]
Comment (hmangla98) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java]: Agreed.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java (hunk 8 above)
Comment (hmangla98) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenarios.java]: Done

**Review by deniskuzZ — APPROVED** [review]
LGTM, pending tests

**Pull request merged**